core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -134,26 +108,120 @@ public static LocalDynamicFilterConsumer create(
+        private final int partitionCount;
+
+        // The resulting predicates from each build-side partition.
+        @GuardedBy("this")
+        private final List<TupleDomain<DynamicFilterId>> partitions;
+
+        public PipelinedExecutionTupleDomainConsumer(int partitionCount)
+        {
+            this.resultFuture = SettableFuture.create();
+            this.partitionCount = partitionCount;
+            this.partitions = new ArrayList<>(partitionCount);
+        }
+
+        @Override
+        public void accept(TupleDomain<DynamicFilterId> tupleDomain)
+        {
+            TupleDomain<DynamicFilterId> result = null;
+            synchronized (this) {
+                // Called concurrently by each DynamicFilterSourceOperator instance (when collection is over).
+                verify(partitions.size() < partitionCount);
+                // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
+                // See the comment at TupleDomain::columnWiseUnion() for more details.
+                partitions.add(tupleDomain);
+                if (partitions.size() == partitionCount || tupleDomain.isAll()) {
+                    // No more partitions are left to be processed.
+                    result = TupleDomain.columnWiseUnion(partitions);
+                }
+            }
+
+            if (result != null) {
+                resultFuture.set(result);
+            }
+        }
+
+        @Override
+        public ListenableFuture<TupleDomain<DynamicFilterId>> getResultFuture()
+        {
+            return resultFuture;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("partitionCount", partitionCount)
+                    .add("partitions", partitions)
+                    .toString();
+        }
+    }
+
+    public static class DynamicFilterSourceTupleDomainConsumer
+            implements TupleDomainConsumer
+    {
+        private final SettableFuture<TupleDomain<DynamicFilterId>> resultFuture;
+        @GuardedBy("this")
+        private TupleDomain<DynamicFilterId> collected = TupleDomain.none();
+
+        public DynamicFilterSourceTupleDomainConsumer()
+        {
+            this.resultFuture = SettableFuture.create();
+        }
+
+        @Override
+        public ListenableFuture<TupleDomain<DynamicFilterId>> getResultFuture()
+        {
+            return resultFuture;
+        }
+
+        @Override
+        public void accept(TupleDomain<DynamicFilterId> tupleDomain)
+        {
+            boolean isCollectedDomainAll;
+            synchronized (this) {
+                collected = TupleDomain.columnWiseUnion(collected, tupleDomain);
+                isCollectedDomainAll = collected.isAll();
+            }
+
+            if (isCollectedDomainAll) {
+                resultFuture.set(TupleDomain.all());
+            }
```

Comment:
Yes, in pipelined mode it was easy to do this lazily. I'm not yet sure how much this needs to be optimised since the thresholds on collection should kick in for large builds. I'm thinking maybe we wait until we have some benchmark results to see if this needs to be improved.